arraycontext/context.py

```diff
@@ -1,4 +1,71 @@
 """
+Freezing and thawing
+--------------------
+
+One of the central concepts introduced by the array context formalism is
+the notion of :meth:`~arraycontext.ArrayContext.freeze` and
+:meth:`~arraycontext.ArrayContext.thaw`. Each array handled by the array context
+is either "thawed" or "frozen". Unlike the real-world concept of freezing and
+thawing, these operations leave the original array alone; instead, a semantically
+separate array in the desired state is returned.
+
+*   "Thawed" arrays are associated with an array context. They use that context
+    to carry out operations (arithmetic, function calls).
+
+*   "Frozen" arrays are static data. They are not associated with an array context,
+    and no operations can be performed on them.
+
+Freezing and thawing may be used to move arrays from one array context to another,
+as long as both array contexts use identical in-memory data representation.
+Otherwise, a common format must be agreed upon, for example using
+:mod:`numpy` through :meth:`~arraycontext.ArrayContext.to_numpy` and
+:meth:`~arraycontext.ArrayContext.from_numpy`.
+Here are some rules of thumb to use when dealing with thawing and freezing:
+
+-   Any array that is stored for a long time needs to be frozen.
+    "Memoized" data (cf. :func:`pytools.memoize` and friends) is a good example
+    of long-lived data that should be frozen.
+
+-   Within a function, if the user did not supply an array context,
+    then any data returned to the user should be frozen. Note that, array contexts
+    need not necessarily be a separate argument.
+
+-   Simply supplying thawed data suffices to provide an array context, using
+    e.g.  :func:`~arraycontext.get_container_context` or
+    :func:`~arraycontext.get_container_context_recursively`; the array context
+    need not be passed in as a separate function argument.
```

Comment:
I like this. A lot. In fact, could be section this out via:
```suggestion
Usage guidelines
^^^^^^^^^^^^^^^^
Here are some rules of thumb to use when dealing with thawing and freezing:

-   Any array that is stored for a long time needs to be frozen.
    "Memoized" data (cf. :func:`pytools.memoize` and friends) is a good example
    of long-lived data that should be frozen.

-   Within a function, if the user did not supply an array context,
    then any data returned to the user should be frozen. Note that, array contexts
    need not necessarily be a separate argument.

-   Simply supplying thawed data suffices to provide an array context, using
    e.g.  :func:`~arraycontext.get_container_context` or
    :func:`~arraycontext.get_container_context_recursively`; the array context
    need not be passed in as a separate function argument.
```

Reply:
Just to make it stick out more